lib/Scheduling/SimplexScheduler.cpp

```diff
@@ -50,9 +50,17 @@ namespace {
 /// side product of solving the parametric problem, and corresponds to the
 /// "RecMII" (= recurrence-constrained minimum II) usually considered as one
 /// component in the lower II bound used by modulo schedulers.
+///
+/// The (acyclic) `Problem` constitutes a special case for this approach; it is
+/// solved by treating it as a `CyclicProblem` in which all dependence distances
+/// are zero.
 class SimplexScheduler {
 private:
-  CyclicProblem &prob;
+  Problem *problem;
+  CyclicProblem *cyclicProblem;
```

Comment:
Obviously not great! These are supposed to point to the same thing, but are actually different pointers because of the virtual inheritance in the `Problem` hierarchy. I tried to get rid of the second member, but none of the "solutions" known to me seem applicable.
- This is not a `static_cast<CyclicProblem*>(problem)`.
- `dynamic_cast<CyclicProblem*>(problem)` is not available because LLVM doesn't use C++'s RTTI.
- I cannot roll out [LLVM's RTTI](https://www.llvm.org/docs/HowToSetUpLLVMStyleRTTI.html) because the problem hierarchy should be extensible by out-of-tree clients, and the ["open"](https://www.llvm.org/docs/HowToSetUpLLVMStyleRTTI.html#rtti-for-open-class-hierarchies) variant doesn't establish direct inheritance between the classes, which I think I need.
- `reinterpret_cast<CyclicProblem*>(problem)` triggers `-Wreinterpret-base-class`.

I could do a `SimplexSchedulerBase` and subclass the algorithm for `Problem` and `CyclicProblem`, but that feels like overkill given that the differences are so small. I think the same is true for a template-based solution.

@teqdruid @mikeurbach, could you help me out here? There must be a better way, but I'm currently just not seeing it 😕

Reply:
Given the goals of this infrastructure, it seems to me that using LLVM's "open" RTTI would be appropriate for the `Problem` hierarchy. I don't really follow why that wouldn't work for your use-case. I think if you went that route, the `SimplexScheduler` would store a `Problem`, and you'd use `llvm::isa<CyclicProblem>(problem)` to perform the necessary checks. Am I missing something?

In general, I'm not a C++ expert, but if the above fanciness isn't working, I think both options you mentioned (using a new base class or using templates) would be fine. I'll try to think of another approach but you've already enumerated the tools I'm familiar with.

Reply:
Thanks, you are right, the open RTTI does work here, I was just using it wrong 🙄. The caveat is that it's limited to single inheritance, and my problem modeling plans currently involve multiple inheritance. A simpler class hierarchy would be intriguing, though -- need to think about that.